flink-runtime/src/test/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManagerTest.java

```diff
@@ -0,0 +1,1092 @@
+		}
+	}
+
+	/**
+	 * Checks that reducing resource requirements while a slot allocation is in-progress results in the slot being freed
+	 * one the allocation completes.
+	 */
+	@Test
+	@Ignore
+	public void testResourceRequirementReductionDuringAllocation() throws Exception {
+		final TaskExecutorConnection taskManagerConnection = createTaskExecutorConnection();
+		final ResourceID resourceID = taskManagerConnection.getResourceID();
+		final SlotID slotId = new SlotID(resourceID, 0);
+
+		final SlotReport slotReport = new SlotReport(createFreeSlotStatus(slotId));
+
+		final DefaultSlotTracker slotTracker = new DefaultSlotTracker();
+		try (DeclarativeSlotManager slotManager = createDeclarativeSlotManagerBuilder()
+			.setSlotTracker(slotTracker)
+			.buildAndStartWithDirectExec()) {
+
+			slotManager.registerTaskManager(taskManagerConnection, slotReport);
+
+			DeclarativeTaskManagerSlot slot = slotTracker.getSlot(slotId);
+
+			final ResourceRequirements resourceRequirements = createResourceRequirementsForSingleSlot();
+			slotManager.processResourceRequirements(resourceRequirements);
+
+			assertSame(SlotState.PENDING, slot.getState());
+
+			slotManager.processResourceRequirements(ResourceRequirements.create(resourceRequirements.getJobId(), "foobar", Collections.emptyList()));
+
+			slot = slotTracker.getSlot(slotId);
+
+			assertSame(SlotState.FREE, slot.getState());
+		}
+	}
+
+	/**
+	 * Tests that pending slot requests are tried to be fulfilled upon new slot registrations.
+	 */
+	@Test
+	public void testFulfillingPendingSlotRequest() throws Exception {
+		final ResourceManagerId resourceManagerId = ResourceManagerId.generate();
+		final ResourceID resourceID = ResourceID.generate();
+		final JobID jobId = new JobID();
+		final SlotID slotId = new SlotID(resourceID, 0);
+		final String targetAddress = "localhost";
+		final ResourceProfile resourceProfile = ResourceProfile.fromResources(42.0, 1337);
+
+		final AtomicInteger numberAllocateResourceCalls = new AtomicInteger(0);
+		ResourceActions resourceManagerActions = new TestingResourceActionsBuilder()
+			.setAllocateResourceConsumer(ignored -> numberAllocateResourceCalls.incrementAndGet())
+			.build();
+
+		final CompletableFuture<Tuple6<SlotID, JobID, AllocationID, ResourceProfile, String, ResourceManagerId>> requestFuture = new CompletableFuture<>();
+		// accept an incoming slot request
+		final TaskExecutorGateway taskExecutorGateway = new TestingTaskExecutorGatewayBuilder()
+			.setRequestSlotFunction(tuple6 -> {
+				requestFuture.complete(Tuple6.of(tuple6.f0, tuple6.f1, tuple6.f2, tuple6.f3, tuple6.f4, tuple6.f5));
+				return CompletableFuture.completedFuture(Acknowledge.get());
+			})
+			.createTestingTaskExecutorGateway();
+
+		final TaskExecutorConnection taskExecutorConnection = new TaskExecutorConnection(resourceID, taskExecutorGateway);
+
+		final SlotStatus slotStatus = new SlotStatus(slotId, resourceProfile);
+		final SlotReport slotReport = new SlotReport(slotStatus);
```

Comment:
This logic looks quite similar to the logic in `testSlotRequestWithFreeSlot `.

Reply:
They are; the only difference is the order in which requirements are declared / the task executor is registered. I'll check whether we can't de-duplicate things a bit.